python/triton/compiler/compiler.py

```diff
@@ -106,28 +90,41 @@ def parse_options(self):
 
 class IRSource:
 
-    def __init__(self, path):
+    def __init__(self, path, context):
         self.path = path
         path = Path(path)
         self.ext = path.suffix[1:]
         self.src = path.read_text()
-        match = re.search(prototype_pattern[self.ext], self.src, re.MULTILINE)
-        self.name = match.group(1)
-        signature = match.group(2)
-        types = re.findall(arg_type_pattern[self.ext], signature)
-        self.signature = {k: convert_type_repr(ty) for k, ty in enumerate(types)}
+        ir.load_dialects(context)
+
+        # We don't have a easy-to-use PTX parser that we can use, so keep that regex for now.
+        # TODO - replace with a proper parser
+        if self.ext == "ptx":
+            match = re.search(prototype_pattern[self.ext], self.src, re.MULTILINE)
+            self.name = match.group(1)
+            signature = match.group(2)
+            types = re.findall(arg_type_pattern[self.ext], signature)
+            self.signature = {k: convert_type_repr(ty) for k, ty in enumerate(types)}
+        else:
+            self.module = ir.parse_mlir_module(self.path, context)
```

Comment:
This implies we parse the full IR even on a cache hit where before we just did a few regex searches. Have you tested how this impacts performance?

Also wondering if `IRSource.signature` is actually used anywhere? Nothing stands out from a quick grep.

Reply:
I believe that this is mainly a debug path (triggered when directly compiling TTIR/TTGIR/PTX files) as opposed to the ASTSource path (source program as input, hot path). The ASTSource path should not be impacted by these changes.

Reply:
Agree with Sam that maintainability > performance for this path. Plus, compiling TTIR / TTGIR is far more expensive than `ir.parse_mlir_module()`.

Reply:
Yes this path is mainly used for tests, but it doesn't hurt to quantify the trade-off we're making here.

> compiling TTIR / TTGIR is far more expensive than ir.parse_mlir_module().

In the cache miss case we obviously need to parse the module anyway so this is irrelevant. This would only matter in the cache hit path where we aren't compiling anything.

Reply:
Can we just store the signature in the cache, and use that in the cache hit path?

Reply:
Thinking about this further... why do we care about performance for the cache hit case? If we are running a bunch of unit tests in CI, we don't expect to hit the cache. If we are doing local perf debugging, then we will warm up the kernel before measuring its runtime.

Reply:
> Have you tested how this impacts latency?

I took a look at this the other day (apologies for the high-latency reply). Over 1k iters `ir.context()` + the added overhead is ~1.1ms on average for the unit test that was added (vector_add ttgir). `triton.compile` is about 1.95ms total (inclusive of the 1.1ms, since on a cache hit we are still grabbing the context + parsing as per the PR). I confirmed on main the overhead of `triton.compile` w/regex is ~0.95ms.

So it's ~2x the latency for the smallest benchmark possible. I suspect for larger benchmarks compilation time would (likely) dominate, and it would be less of an issue.

That being said, as mentioned by @embg, the only reason we have to parse the MLIR in the cache hit case seems to be for:

 `key = f"{triton_key()}-{src.hash()}-{backend.hash()}-{options.hash()}-{str(sorted(env_vars.items()))}"`
 
 which depends on `src.parse_options()` (obtained from parsing the MLIR/regex previously).
 I'm actually not sure why this needs to be a part of the hash, since as far as I can tell all it does is grab `triton_gpu.num-warps`---which should already be in `src.hash()` (since it's embedded in the TTGIR/TTIR text file).
 
 If we remove `src.parse_options()` from the hash we can push the parsing down to only affect the cache miss path---in which case there should be zero added overhead for the cache hit (and we can get rid of the regex).

Reply:
>  it's ~2x the latency for the smallest benchmark possible.

Given that this isn't a performance critical path I think this is probably okay then. I'm assuming most of the added latency comes from context and dialect initialisation so it won't scale terribly with larger IR files. 

cc @ThomasRaoux for a second opinion though

Reply:
sorry missed this conversation. I agree since this is a debug path I don't think we should worry that much about launch performance